### PR TITLE
Revamp class file structure

### PR DESCRIPTION
**Pull request by malleoz**
Classes are now organized in a mkw_classes folder. Classes can be individually imported by doing mkw_classes.ClassName. Each class allows construction based off an address, or (in cases where there is only one way to determine the class instance's reference) by a player index for example.

This should allow for more readable navigation of class hierarchies and easier debugging for future script work.

I've tested all existing scripts in this repository to assure no behavior is altered, however an additional thorough review for InfoDisplay would be appreciated.

### CODE REVIEW

File: scripts/MKW_TTK.py [hunk 1]
```diff
@@ -20,14 +22,15 @@ def onStateLoad(is_slot, slot):
...
+    in_race = RaceManager.state().value >= RaceState.COUNTDOWN.value
+
     ghostInput = ghostInputs[frame]
-    if (ghostInput and classes.RaceInfo.stage() >= 1):
+    if (ghostInput and in_race):
         TTK_Lib.writeGhostInputs(ghostInput)
     
     playerInput = playerInputs[frame]
-    if (playerInput and classes.RaceInfo.stage() >= 1):
+    if (playerInput and in_race):
```
Comment (vabold) [inline scripts/MKW_TTK.py:33]: > Applying suggestions on deleted lines is not supported.

Thanks GitHub. Minor nitpick, but we should call this `input_ready` and only allow states 1 and 2.

File: scripts/MKW_TTK_GhostOnly.py [hunk 2]
```diff
@@ -18,10 +20,11 @@ def onStateLoad(is_slot, slot):
 @event.on_frameadvance
 def onFrameAdvance():
     global ghostInputs
-    frame = core.get_frame_of_input()
+    frame = frame_of_input()
+    in_race = RaceManager.state().value >= RaceState.COUNTDOWN.value
     
     ghostInput = ghostInputs[frame]
-    if (ghostInput and classes.RaceInfo.stage() >= 1):
+    if (ghostInput and in_race):
```
Comment (vabold) [inline scripts/MKW_TTK_GhostOnly.py:27]: Same here.

File: scripts/MKW_TTK_PlayerOnly.py [hunk 3]
```diff
@@ -18,10 +20,11 @@ def onStateLoad(is_slot, slot):
 @event.on_frameadvance
 def onFrameAdvance():
     global playerInputs
-    frame = core.get_frame_of_input()
+    frame = frame_of_input()
+    in_race = RaceManager.state().value >= RaceState.COUNTDOWN.value
     
     playerInput = playerInputs[frame]
-    if (playerInput and classes.RaceInfo.stage() >= 1):
+    if (playerInput and in_race):
```
Comment (vabold) [inline scripts/MKW_TTK_PlayerOnly.py:27]: Same here.

File: scripts/Modules/TTK_Lib.py [hunk 4]
```diff
@@ -48,31 +55,40 @@ def encodeTrickInput(input):
...
+        stream_addr = ghost_writer.button_stream(inputType.value)
+        button_stream = GhostButtonsStream(addr=stream_addr)
+        address = button_stream.buffer()
+        endAddrOffset = button_stream.size()
     else:
-        addresses = classes.getGhostAddresses()
-        endAddrOffset = memory.read_u32(classes.getGhostAddressLengthPointer()[2])
+        # TODO: Ghost is index=1 if you are racing a ghost, but if you watch a replay
+        # for example, the ghost is index 0. We want to support this scenario, so
+        # we'll need to determine how to set controllerIdx appropriately.
```
Comment (vabold) [inline scripts/Modules/TTK_Lib.py:76]: The answer is to evaluate the current gamemode in RaceConfig.

File: scripts/Modules/TTK_Lib.py [hunk 5]
```diff
@@ -382,19 +407,21 @@ def controller_patch() -> None:
...
-            memory.read_u32(classes.InputMgr.chain()) + 0x8)
-    set_buttons(inputs, player_controller)
+    kart_input = KartInput(addr=PlayerInput.kart_input(playerIdx=0))
+    controller = Controller(addr=kart_input.race_controller())
+    set_buttons(inputs, controller)
 
-def set_buttons(inputs, controller):
-    memory.write_u8(controller + 0x4d, inputs.accel + (inputs.brake << 1) +
+def set_buttons(inputs, controller : Controller):
+    """This writes button data to addresses with implicit padding in structs"""
```
Comment (vabold) [inline scripts/Modules/TTK_Lib.py:421]: I would also recommend referencing where the new data is used in the docstring.

File: scripts/Modules/mkw_classes/common.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        return vec3(self.x + other.x, self.y + other.y, self.z + other.z)
+
+    def __sub__(self, other):
+        return vec3(self.x - other.x, self.y - other.y, self.z - other.z)
+
+    def norm_xyz(self) -> float:
+        return math.sqrt(self.x**2 + self.y**2 + self.z**2)
+
+    def norm_xz(self) -> float:
+        return math.sqrt(self.x**2 + self.z**2)
```
Comment (vabold) [inline scripts/Modules/mkw_classes/common.py:45]:
```suggestion
    def length(self) -> float:
        return math.sqrt(self.x**2 + self.y**2 + self.z**2)

    def length_xz(self) -> float:
        return math.sqrt(self.x**2 + self.z**2)
```
`length` is a known `EGG::Vector3f` symbol.

**Review by vabold — CHANGES_REQUESTED** [review]
This is a monster PR so reviewing this in-depth will take some time.

**Comment by vabold** [comment]
Known issues out of scope of this PR:

* Full EGG parity with `vec2`/`vec3`/`mat34`/`quatf`
* Proper `WheelCount` documentation
* Broadening TTK support by factoring in current gamemode

As well as this, we may want to introduce broader documentation for the player indices (and possibly the input system), as this will be helpful to other developers.

File: scripts/Modules/TTK_Lib.py (hunk 4 above)
Comment (malleoz) [reply scripts/Modules/TTK_Lib.py:76]: See https://github.com/TASLabz/mkw-scripts/issues/42

**Comment by malleoz** [comment]
Applied all of the above suggestions.

**Review by vabold — APPROVED** [review]

**Pull request merged**